File: leetcode/max_number_of_k-sum_pairs/max_number_of_k-sum_pairs.cpp
class Solution {
public:
    int maxOperations(vector<int>& nums, int k) {
        unordered_map<int,int> cache;

        int ans=0;
        for(int i=0;i<nums.size();i++){
            int num=nums[i];
            if(k-num<=0){
                continue;
            }
            auto iter = cache.find(k-num);
            if(iter==cache.end()){
                auto num_cache=cache.find(num);
                if(num_cache!=cache.end()){
                    (*num_cache).second++;
                }else{
                    cache.insert({num,1});
                }
                continue;
            }
            if((*iter).second!=1){
                (*iter).second--;
            }else{
                cache.erase(iter);
            }
            ans++;
        }

        return ans;
    }
};

File: leetcode/max_number_of_k-sum_pairs/max_number_of_k-sum_pairs.py
class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        all_dict=defaultdict(int)
        ans=0
        for num in nums:
            if all_dict[k-num]!=0:
                ans+=1
                all_dict[k-num]-=1
            else:
                all_dict[num]+=1

        return ans

